### PR TITLE
Proposal for improved Helm-Manifest-Deployer

### PR DESCRIPTION
**Pull request by achimweigel**
**How to categorize this PR?**

/area deployers
/kind discussion
/priority 3

**What this PR does / why we need it**:

Proposal for improved Helm-Manifest-Deployer.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
- proposal for improved Helm-Manifest-Deployer
```

### CODE REVIEW

File: docs/proposals/helm-manifest-deployer-ordering.md [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+The basic order of deleting the deployed manifests remains the same as before and is divided into three deletion groups:
+
+- Namespaced objects deployed by the Chart
+- Not namespaced objects deployed by the Chart (except CRDs)
+- CRDs deployed by the Chart
+
+The deletion continues only with the next object/deletion group if all objects from the groups before are gone. This is 
+different to the current approach. The deletion is tried as long until all objects of all 3 deletion groups are gone
+or the specified configurable timeout (default 5 min.) comes into the game and the deletion failed.
```
Comment (reshnm) [inline docs/proposals/helm-manifest-deployer-ordering.md:52]:
```suggestion
or the specified configurable timeout (default 5 min.) is expired and the deletion failed.
```

File: docs/proposals/helm-manifest-deployer-ordering.md [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+      apiVersion: helm.deployer.landscaper.gardener.cloud/v1alpha1
+      kind: ProviderConfiguration
+      ...
+      deletionGroups:
+        - predefinedResourceGroup: 
+            type: namespaced-resources
+        - predefinedResourceGroup:
+            type: cluster-scoped-resources     # does not include the crds
+        - predefinedResourceGroup: 
+            type: crds
```
Comment (reshnm) [inline docs/proposals/helm-manifest-deployer-ordering.md:100]:
```suggestion
      deletionGroups:
        predefinedResourceGroups: 
          - type: namespaced-resources
          - type: cluster-scoped-resources     # does not include the crds
          - type: crds
```

Wouldn't this beaser to write and parse (as a human :-) )?

File: docs/proposals/helm-manifest-deployer-ordering.md [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+##### Example: skip deletion of CRDs
+
+In this example, the deletion of CRDs is skipped:
+
+```yaml
+deletionGroups:
+  - predefinedResourceGroup: 
+      type: namespaced-resources
+  - predefinedResourceGroup: 
+      type: cluster-scoped-resources
```
Comment (reshnm) [inline docs/proposals/helm-manifest-deployer-ordering.md:116]:
```suggestion
deletionGroups:
  predefinedResourceGroups: 
    - type: namespaced-resources
    - type: cluster-scoped-resources
```

File: docs/proposals/helm-manifest-deployer-ordering.md [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+```yaml
+deletionGroups:
+  - resources:
+      - group:   ...
+        version: ...
+        kind:    ...
+      - group:   ...
+        version: ...
+        kind:    ...
```
Comment (reshnm) [inline docs/proposals/helm-manifest-deployer-ordering.md:132]:
```suggestion
deletionGroups:
  resources:
    - group:   ...
       version: ...
       kind:    ...
    - group:   ...
        version: ...
        kind:    ...
```

File: docs/proposals/helm-manifest-deployer-ordering.md [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        kind:    "configmaps"
+      - group:   ""
+        version: "v1"
+        kind:    "secrets"
+  - predefinedResourceGroup: 
+      type: namespaced-resources
+  - predefinedResourceGroup: 
+      type: cluster-scoped-resources
+  - predefinedResourceGroup: 
+      type: crds
```
Comment (reshnm) [inline docs/proposals/helm-manifest-deployer-ordering.md:155]:
```suggestion
deletionGroups:
  resources:
    - group:   ""
       version: "v1"
       kind:    "configmaps"
    - group:   ""
       version: "v1"
       kind:    "secrets"
  predefinedResourceGroups: 
    - type: namespaced-resources
    - type: cluster-scoped-resources
    - type: crds
```

**Review by reshnm — CHANGES_REQUESTED** [review]
some ideas

File: docs/proposals/helm-manifest-deployer-ordering.md (hunk 2 above)
Comment (achimweigel) [reply docs/proposals/helm-manifest-deployer-ordering.md]: resolved as discussed offline

**Review by robertgraeff — APPROVED** [review]

**Pull request merged**